File: apps/mobile/lib/features/media_upload/application/upload_queue.dart
import 'dart:io';
import 'dart:math' as math;

import 'package:crypto/crypto.dart';
import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../../core/network/dio_client.dart';
import '../data/image_compressor.dart';
import '../data/media_upload_transport.dart';
import '../data/upload_queue_store.dart';
import '../domain/upload_item.dart';

/// The photos of one post (or store, or chat) being prepared: compressed on
/// the phone, uploaded straight to storage, confirmed with the API.
///
///  - Two uploads run at once; the rest wait their turn.
///  - A failed step is retried with backoff when it's worth retrying (network,
///    timeout, rate limit); a refused file (not an image, too big) fails at once.
///  - Every state change is saved, so a queue killed mid-upload resumes on the
///    next launch ([restore]): unfinished items start again from the step they
///    need (a presigned URL expires, so an interrupted PUT is redone whole —
///    cheap, since files are ~150 KB after compression).
///  - Items can be reordered and removed at any time; removing one cancels its
///    upload. The server deletes anything uploaded but never attached (orphan
///    sweep), so nothing needs to be "un-uploaded".
class UploadQueue extends ChangeNotifier {
  UploadQueue({
    required this.queueId,
    required this._compressor,
    required this._transport,
    required this._store,
    this.maxItems = defaultMaxItems,
    this.concurrency = defaultConcurrency,
    this.retryDelays = defaultRetryDelays,
    this._delay = Future<void>.delayed,
    String Function()? newId,
  }) : _newId = newId ?? _randomId;

  /// Photos per post.
  static const defaultMaxItems = 10;
  static const defaultConcurrency = 2;

  /// Waits before the 2nd, 3rd and 4th attempt; then the item fails.
  static const defaultRetryDelays = [
    Duration(seconds: 2),
    Duration(seconds: 6),
    Duration(seconds: 20),
  ];

  final String queueId;
  final int maxItems;
  final int concurrency;
  final List<Duration> retryDelays;
  final ImageCompressor _compressor;
  final MediaUploadTransport _transport;
  final UploadQueueStore _store;
  final Future<void> Function(Duration) _delay;
  final String Function() _newId;

  List<UploadItem> _items = [];
  final Map<String, CancelToken> _cancelTokens = {};
  final Set<String> _running = {};
  bool _disposed = false;

  List<UploadItem> get items => List.unmodifiable(_items);
  int get remainingSlots => maxItems - _items.length;

  /// True when every photo is uploaded and confirmed.
  bool get isComplete =>
      _items.isNotEmpty && _items.every((i) => i.status == UploadStatus.done);

  /// Confirmed media ids in the user's chosen order, for the post.
  List<String> get mediaIds => [
    for (final item in _items)
      if (item.status == UploadStatus.done) item.mediaId!,
  ];

  /// Loads a saved queue and resumes whatever was unfinished.
  Future<void> restore() async {
    final saved = await _store.load(queueId);
    _items = [
      for (final item in saved)
        item.isActive
            ? item.copyWith(status: UploadStatus.queued, progress: 0)
            : item,
    ];
    _changed();
    _pump();
  }

  /// Adds picked photos, up to [maxItems] in total; returns how many were taken.
  Future<int> add(List<String> sourcePaths) async {
    final taken = sourcePaths.take(math.max(0, remainingSlots)).toList();
    _items = [
      ..._items,
      for (final path in taken) UploadItem(id: _newId(), sourcePath: path),
    ];
    await _save();
    _pump();
    return taken.length;
  }

  Future<void> remove(String id) async {
    final item = _find(id);
    if (item == null) return;
    _cancelTokens.remove(id)?.cancel();
    _items = [
      for (final i in _items)
        if (i.id != id) i,
    ];
    await _save();
    await _store.deleteFile(item.compressedPath);
    _pump();
  }

  /// Moves an item so it ends up at [newIndex] (`onReorderItem` semantics).
  Future<void> reorder(int oldIndex, int newIndex) async {
    final items = [..._items];
    final item = items.removeAt(oldIndex);
    items.insert(newIndex, item);
    _items = items;
    await _save();
  }

  Future<void> retry(String id) async {
    final item = _find(id);
    if (item == null || item.status != UploadStatus.failed) return;
    _update(
      id,
      (i) => i.copyWith(
        status: UploadStatus.queued,
        attempts: 0,
        progress: 0,
        clearError: true,
      ),
    );
    await _save();
    _pump();
  }

  @override
  void dispose() {
    _disposed = true;
    for (final token in _cancelTokens.values) {
      token.cancel();
    }
    super.dispose();
  }

  void _pump() {
    if (_disposed) return;
    for (final item in _items) {
      if (_running.length >= concurrency) break;
      if (item.status == UploadStatus.queued && !_running.contains(item.id)) {
        _running.add(item.id);
        _run(item.id).whenComplete(() {
          _running.remove(item.id);
          _pump();
        });
      }
    }
  }

  Future<void> _run(String id) async {
    while (!_disposed && _find(id) != null) {
      try {
        await _attempt(id);
        return;
      } on UploadFailure catch (failure) {
        if (_find(id) == null || failure.code == 'cancelled') return;
        final attempts = _find(id)!.attempts + 1;
        final retry = failure.retryable && attempts <= retryDelays.length;
        _update(
          id,
          (i) => i.copyWith(
            attempts: attempts,
            status: retry ? UploadStatus.queued : UploadStatus.failed,
            errorCode: failure.code,
            progress: 0,
          ),
        );
        await _save();
        if (!retry) return;
        await _delay(retryDelays[attempts - 1]);
      } on Exception {
        // Compression or file errors: the photo can't be used as it is.
        if (_find(id) == null) return;
        _update(
          id,
          (i) =>
              i.copyWith(status: UploadStatus.failed, errorCode: 'compression'),
        );
        await _save();
        return;
      }
    }
  }

  Future<void> _attempt(String id) async {
    var item = _find(id)!;
    final compressed = item.compressedPath;
    if (compressed == null || !await File(compressed).exists()) {
      _update(id, (i) => i.copyWith(status: UploadStatus.compressing));
      final target = await _store.compressedPathFor(id);
      final size = await _compressor.compress(item.sourcePath, target);
      if (_find(id) == null) return _store.deleteFile(target);
      _update(id, (i) => i.copyWith(compressedPath: target, byteSize: size));
      await _save();
    }

    item = _find(id)!;
    final file = File(item.compressedPath!);
    final bytes = await file.readAsBytes();
    _update(id, (i) => i.copyWith(status: UploadStatus.uploading, progress: 0));
    await _save();

    final target = await _transport.presign(
      byteSize: bytes.length,
      sha256: sha256.convert(bytes).toString(),
      contentType: 'image/webp',
    );
    final token = _cancelTokens[id] = CancelToken();
    try {
      await _transport.put(
        target,
        file,
        cancelToken: token,
        onProgress: (progress) =>
            _update(id, (i) => i.copyWith(progress: progress)),
      );
    } finally {
      _cancelTokens.remove(id);
    }
    if (_find(id) == null) return;

    _update(
      id,
      (i) => i.copyWith(status: UploadStatus.confirming, progress: 1),
    );
    await _transport.confirm(target.mediaId);
    _update(
      id,
      (i) => i.copyWith(
        status: UploadStatus.done,
        mediaId: target.mediaId,
        clearError: true,
      ),
    );
    await _save();
  }

  UploadItem? _find(String id) => _items.where((i) => i.id == id).firstOrNull;

  /// Changes one item in memory; callers [_save] at the state changes worth persisting.
  void _update(String id, UploadItem Function(UploadItem) change) {
    _items = [
      for (final i in _items)
        if (i.id == id) change(i) else i,
    ];
    _changed();
  }

  Future<void> _save() async {
    await _store.save(queueId, _items);
    _changed();
  }

  void _changed() {
    if (!_disposed) notifyListeners();
  }

  static String _randomId() {
    final random = math.Random.secure();
    return List.generate(
      16,
      (_) => random.nextInt(256).toRadixString(16).padLeft(2, '0'),
    ).join();
  }
}

/// One queue per draft (e.g. `post-draft:<uuid>`); restored on first use.
final uploadQueueProvider = Provider.autoDispose.family<UploadQueue, String>((
  ref,
  queueId,
) {
  final queue = UploadQueue(
    queueId: queueId,
    compressor: PluginImageCompressor(),
    transport: DioMediaUploadTransport(ref.watch(dioClientProvider)),
    store: PersistentUploadQueueStore(),
  );
  ref.onDispose(queue.dispose);
  queue.restore();
  return queue;
});

File: apps/web/src/lib/media/upload-queue.ts
import { UploadFailure, type UploadTransport } from './upload-transport';

export type UploadStatus =
  'queued' | 'compressing' | 'uploading' | 'confirming' | 'done' | 'failed';

export interface UploadItem {
  readonly id: string;
  /** The picked file, kept for its preview and in case compression must rerun. */
  readonly file: Blob;
  readonly status: UploadStatus;
  /** 0–1 while uploading. */
  readonly progress: number;
  readonly attempts: number;
  readonly errorCode?: string | undefined;
  readonly mediaId?: string;
  readonly compressed?: Blob;
}

export interface UploadQueueOptions {
  transport: UploadTransport;
  compress: (file: Blob) => Promise<Blob>;
  maxItems?: number;
  concurrency?: number;
  /** Waits before the 2nd, 3rd and 4th attempt; then the item fails. */
  retryDelaysMs?: readonly number[];
  sleep?: (ms: number) => Promise<void>;
  newId?: () => string;
}

export const DEFAULT_MAX_ITEMS = 10;
const DEFAULT_CONCURRENCY = 2;
export const DEFAULT_RETRY_DELAYS_MS = [2_000, 6_000, 20_000] as const;

/**
 * The photos being prepared for one post — the web twin of the mobile app's
 * UploadQueue, minus persistence (a closed tab takes its File handles with it):
 *
 *  - compress → hash → presign → PUT to storage with progress → confirm;
 *  - two at a time; the rest wait their turn;
 *  - retryable failures (network, timeout, 5xx, rate limit) back off and try
 *    again; a refused file fails at once and can be retried by hand;
 *  - reorder and remove at any time; removing aborts that item's upload. The
 *    server sweeps uploads that are never attached, so nothing is undone.
 *
 * Framework-free, with `subscribe`/`getSnapshot` for useSyncExternalStore.
 */
export class UploadQueue {
  readonly maxItems: number;
  private readonly concurrency: number;
  private readonly retryDelaysMs: readonly number[];
  private readonly sleep: (ms: number) => Promise<void>;
  private readonly newId: () => string;

  private items: readonly UploadItem[] = [];
  private readonly listeners = new Set<() => void>();
  private readonly controllers = new Map<string, AbortController>();
  private readonly running = new Set<string>();
  private disposed = false;

  constructor(private readonly options: UploadQueueOptions) {
    this.maxItems = options.maxItems ?? DEFAULT_MAX_ITEMS;
    this.concurrency = options.concurrency ?? DEFAULT_CONCURRENCY;
    this.retryDelaysMs = options.retryDelaysMs ?? DEFAULT_RETRY_DELAYS_MS;
    this.sleep = options.sleep ?? ((ms) => new Promise((r) => setTimeout(r, ms)));
    this.newId = options.newId ?? (() => crypto.randomUUID());
  }

  subscribe = (listener: () => void): (() => void) => {
    this.listeners.add(listener);
    return () => this.listeners.delete(listener);
  };

  getSnapshot = (): readonly UploadItem[] => this.items;

  get remainingSlots(): number {
    return Math.max(0, this.maxItems - this.items.length);
  }

  /** Confirmed media ids in the user's order — the first is the cover. */
  get mediaIds(): string[] {
    return this.items.flatMap((i) => (i.status === 'done' && i.mediaId ? [i.mediaId] : []));
  }

  get isComplete(): boolean {
    return this.items.length > 0 && this.items.every((i) => i.status === 'done');
  }

  /** Adds files up to the limit; returns how many were taken. */
  add(files: readonly Blob[]): number {
    const taken = files.slice(0, this.remainingSlots);
    this.set([
      ...this.items,
      ...taken.map((file) => ({
        id: this.newId(),
        file,
        status: 'queued' as const,
        progress: 0,
        attempts: 0,
      })),
    ]);
    this.pump();
    return taken.length;
  }

  remove(id: string): void {
    this.controllers.get(id)?.abort();
    this.set(this.items.filter((i) => i.id !== id));
    this.pump();
  }

  /** Moves an item so it ends up at `to`. */
  reorder(from: number, to: number): void {
    if (from === to || from < 0 || to < 0 || from >= this.items.length || to >= this.items.length) {
      return;
    }
    const items = [...this.items];
    const [item] = items.splice(from, 1);
    items.splice(to, 0, item!);
    this.set(items);
  }

  retry(id: string): void {
    if (this.find(id)?.status !== 'failed') return;
    this.update(id, { status: 'queued', attempts: 0, progress: 0, errorCode: undefined });
    this.pump();
  }

  dispose(): void {
    this.disposed = true;
    for (const controller of this.controllers.values()) controller.abort();
    this.listeners.clear();
  }

  private pump(): void {
    if (this.disposed) return;
    for (const item of this.items) {
      if (this.running.size >= this.concurrency) break;
      if (item.status === 'queued' && !this.running.has(item.id)) {
        this.running.add(item.id);
        void this.run(item.id).finally(() => {
          this.running.delete(item.id);
          this.pump();
        });
      }
    }
  }

  private async run(id: string): Promise<void> {
    while (!this.disposed && this.find(id)) {
      try {
        await this.attempt(id);
        return;
      } catch (error) {
        const item = this.find(id);
        if (!item) return;
        const failure =
          error instanceof UploadFailure ? error : new UploadFailure('compression', false);
        if (failure.code === 'cancelled') return;
        const attempts = item.attempts + 1;
        const retry = failure.retryable && attempts <= this.retryDelaysMs.length;
        this.update(id, {
          attempts,
          status: retry ? 'queued' : 'failed',
          errorCode: failure.code,
          progress: 0,
        });
        if (!retry) return;
        await this.sleep(this.retryDelaysMs[attempts - 1]!);
      }
    }
  }

  private async attempt(id: string): Promise<void> {
    let item = this.find(id)!;
    if (!item.compressed) {
      this.update(id, { status: 'compressing' });
      const compressed = await this.options.compress(item.file);
      if (!this.find(id)) return;
      this.update(id, { compressed });
      item = this.find(id)!;
    }
    const body = item.compressed!;

    this.update(id, { status: 'uploading', progress: 0 });
    const target = await this.options.transport.presign({
      byteSize: body.size,
      sha256: await sha256Hex(body),
      contentType: body.type,
    });
    if (!this.find(id)) return;

    const controller = new AbortController();
    this.controllers.set(id, controller);
    try {
      await this.options.transport.put(
        target,
        body,
        (progress) => this.update(id, { progress }),
        controller.signal,
      );
    } finally {
      this.controllers.delete(id);
    }
    if (!this.find(id)) return;

    this.update(id, { status: 'confirming', progress: 1 });
    await this.options.transport.confirm(target.mediaId);
    this.update(id, { status: 'done', mediaId: target.mediaId, errorCode: undefined });
  }

  private find(id: string): UploadItem | undefined {
    return this.items.find((i) => i.id === id);
  }

  private update(id: string, change: Partial<UploadItem>): void {
    if (!this.find(id)) return;
    this.set(this.items.map((i) => (i.id === id ? { ...i, ...change } : i)));
  }

  private set(items: readonly UploadItem[]): void {
    this.items = items;
    for (const listener of this.listeners) listener();
  }
}

export async function sha256Hex(blob: Blob): Promise<string> {
  const digest = await crypto.subtle.digest('SHA-256', await blob.arrayBuffer());
  return Array.from(new Uint8Array(digest), (b) => b.toString(16).padStart(2, '0')).join('');
}
